src/topology/category/Compactum.lean

```diff
@@ -0,0 +1,467 @@
+instance : has_coe_to_sort Compactum := ⟨Type*,forget.obj⟩
+instance {X Y : Compactum} : has_coe_to_fun (X ⟶ Y) := ⟨λ f, X → Y, λ f, f.f⟩
+instance : has_limits Compactum := has_limits_of_has_limits_creates_limits forget
+
+/-- The structure map for a compactum, essentially sending an ultrafilter to its limit. -/
+def str (X : Compactum) : ultrafilter X → X := X.a
+
+/-- The monadic join. -/
+def join (X : Compactum) : ultrafilter (ultrafilter X) → ultrafilter X := (μ_ β).app _
+
+/-- The inclusion of `X` into `ultrafilter X`. -/
+def incl (X : Compactum) : X → ultrafilter X := (η_ β).app _
+
+@[simp] lemma str_incl (X : Compactum) (x : X) : X.str (X.incl x) = x :=
+begin
+  change ((η_ β).app _ ≫ X.a) _ = _,
+  rw monad.algebra.unit,
+  refl,
+end
+
+@[simp] lemma str_hom_commute (X Y : Compactum) (f : X ⟶ Y) (xs : ultrafilter X) :
+  f (X.str xs) = Y.str (ultrafilter.map f xs) :=
+begin
+  change (X.a ≫ f.f) _ = _,
+  rw ←f.h,
+  refl,
+end
+
+@[simp] lemma join_distrib (X : Compactum) (uux : ultrafilter (ultrafilter X)) :
+  X.str (X.join uux) = X.str (ultrafilter.map X.str uux) :=
+begin
+  change ((μ_ β).app _ ≫ X.a) _ = _,
+  rw monad.algebra.assoc,
+  refl,
+end
+
+instance {X : Compactum} : topological_space X :=
+{ is_open := { U | ∀ (F : ultrafilter X), X.str F ∈ U → U ∈ F.1 },
+  is_open_univ := λ _ _, filter.univ_sets _,
+  is_open_inter := λ S T h3 h4 h5 h6,
+    filter.inter_sets _ (h3 _ h6.1) (h4 _ h6.2),
+  is_open_sUnion := λ S h1 F ⟨T,hT,h2⟩,
+    filter.sets_of_superset _ (h1 T hT _ h2) (set.subset_sUnion_of_mem hT) }
+
+theorem is_closed_iff {X : Compactum} (S : set X) : is_closed S ↔
+  (∀ F : ultrafilter X, S ∈ F.1 → X.str F ∈ S) :=
+begin
+  split,
+  { intros cond F h,
+    by_contradiction c,
+    specialize cond F c,
+    cases F with F hF,
+    rw ultrafilter_iff_compl_mem_iff_not_mem at hF,
+    rw hF at cond,
+    contradiction },
+  { intros h1 F h2,
+    specialize h1 F,
+    cases mem_or_compl_mem_of_ultrafilter F.2 S;
+    finish }
+end
+
+instance {X : Compactum} : compact_space X :=
+begin
+  constructor,
+  rw compact_iff_ultrafilter_le_nhds,
+  intros F hF h,
+  refine ⟨X.str ⟨F,hF⟩, by tauto, _⟩,
+  rw le_nhds_iff,
+  intros S h1 h2,
+  exact h2 ⟨F,hF⟩ h1,
+end
+
+/-- A local definition used only in the proofs. -/
+private def basic {X : Compactum} (A : set X) : set (ultrafilter X) := {F | A ∈ F.1}
+/-- A local definition used only in the proofs. -/
+private def cl {X : Compactum} (A : set X) : set X := X.str '' (basic A)
+
+private lemma basic_inter {X : Compactum} (A B : set X) : basic (A ∩ B) = basic A ∩ basic B :=
+begin
+  ext G,
+  split,
+  { intro hG,
+    refine ⟨_, _⟩;
+    apply filter.sets_of_superset _ hG;
+    intros x hx;
+    finish },
+  { rintros ⟨h1,h2⟩,
+    exact G.val.inter_sets h1 h2 }
+end
+
+private lemma subset_cl {X : Compactum} (A : set X) : A ⊆ cl A := λ a ha, ⟨X.incl a, ha,by simp⟩
+
+private theorem cl_cl {X : Compactum} (A : set X) : cl (cl A) ⊆ cl A :=
```

Comment:
Is this long private theorem subsumed by some later public statement? If not then it shouldn't be private.

Reply:
I added two lemmas:
1. `Lim_eq_str` saying that the limit of an ultrafilter is the same as its image under the structure map.
2. `cl_eq_closure` saying that `cl` is the same as the closure of a set.
The second one reduces this `cl_cl` theorem to a known theorem about closures.